src/adapter/src/catalog.rs

```diff
@@ -6364,6 +6479,19 @@ impl SessionCatalog for ConnCatalog<'_> {
         &self.state.roles_by_id[id]
     }
 
+    fn collect_role_membership(&self, id: &RoleId) -> BTreeSet<RoleId> {
+        let mut membership = BTreeSet::new();
+        let mut queue = VecDeque::from(vec![id]);
+        while let Some(id) = queue.pop_front() {
+            if !membership.contains(id) {
+                membership.insert(id.clone());
+                let role = self.get_role(id);
```

Comment:
nit: thinking about detecting bugs in the future, it could be nice to error loudly here (e.g. `tracing::error!` or a `soft_assert!`) if we find a circular reference

Reply:
I added a `soft_assert!` here.